### PR TITLE
Add a user friendly error message when no development team is selected for physical deployment

### PR DESCRIPTION
**Pull request by xster**

### CODE REVIEW

**Comment by xster** [comment]
For #7839

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 1]
```diff
@@ -206,15 +213,57 @@ void diagnoseXcodeBuildFailure(XcodeBuildResult result) {
...
+      && result.buildCommands != null
+      && result.buildCommands.contains('build')
+      && result.appDirectory != null) {
+    RunResult checkBuildSettings = await runAsync(
+      result.buildCommands..add('-showBuildSettings'),
+      workingDirectory: result.appDirectory,
+      allowReentrantFlutter: true
+    );
+    if (checkBuildSettings.exitCode == 0
+        && !checkBuildSettings.stdout?.contains('DEVELOPMENT_TEAM') == true) {
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/ios/mac.dart]: This is definitely the case with recent releases of Xcode (> 8.1, IIRC) when using automatic signing -- is it also the case in Xcode 7 and/or when you use manual mode and use a provisioning profile? 

One option would be to bump the minimum required Xcode version from its current value (IIRC, >= 7.0) in `mac.dart`.

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 2]
```diff
@@ -206,15 +213,57 @@ void diagnoseXcodeBuildFailure(XcodeBuildResult result) {
...
+      printError('  1- Opening the Flutter project\'s Xcode target with');
+      printError('       open ios/Runner.xcodeproj');
+      printError("  2- Select the 'Runner' project in the navigator then the 'Runner' target");
+      printError('     in the project settings');
+      printError("  3- In the 'General' tab, make sure a 'Development Team' is selected\n");
+      printError('For more information, please visit:');
+      printError('  https://flutter.io/setup/#deploy-to-ios-devices\n');
+      printError('Or run on an iOS simulator');
+      printError(
+        '═══════════════════════════════════════════════════════════════════════════════════');
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Consider using a multi-line `'''` string.

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 3]
```diff
@@ -206,15 +213,57 @@ void diagnoseXcodeBuildFailure(XcodeBuildResult result) {
     printError("  open ios/Runner.xcodeproj");
     return;
   }
+  if (result.buildForDevice
+      && result.buildCommands != null
+      && result.buildCommands.contains('build')
+      && result.appDirectory != null) {
```
Comment (cbracken) [inline packages/flutter_tools/lib/src/ios/mac.dart]: When buildForDevice is true, are buildCommands and appDirectory always expected? If so, use assertions ahead of the conditional.

File: packages/flutter_tools/lib/src/commands/build_ios.dart [hunk 4]
```diff
@@ -77,7 +77,7 @@ class BuildIOSCommand extends BuildSubCommand {
 
     if (!result.success) {
       printError('Encountered error while building for $logTarget.');
-      diagnoseXcodeBuildFailure(result);
+      await diagnoseXcodeBuildFailure(result);
       throwToolExit('');
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: It seems hacky to call it with an empty string.

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 5]
```diff
@@ -169,7 +169,14 @@ Future<XcodeBuildResult> buildXcodeProject({
       printStatus(result.stderr);
     if (result.stdout.isNotEmpty)
       printStatus(result.stdout);
-    return new XcodeBuildResult(false, stdout: result.stdout, stderr: result.stderr);
+    return new XcodeBuildResult(
+      false,
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/ios/mac.dart]: we should use named arguments for all boolean parameters.

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 6]
```diff
@@ -206,15 +213,57 @@ void diagnoseXcodeBuildFailure(XcodeBuildResult result) {
     printError("  open ios/Runner.xcodeproj");
     return;
   }
+  if (result.buildForDevice
+      && result.buildCommands != null
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/ios/mac.dart]: if you put the && at the end of the lines you'll get the subexpressions lined up

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 7]
```diff
@@ -206,15 +213,57 @@ void diagnoseXcodeBuildFailure(XcodeBuildResult result) {
...
+      && result.appDirectory != null) {
+    RunResult checkBuildSettings = await runAsync(
+      result.buildCommands..add('-showBuildSettings'),
+      workingDirectory: result.appDirectory,
+      allowReentrantFlutter: true
+    );
+    if (checkBuildSettings.exitCode == 0
+        && !checkBuildSettings.stdout?.contains('DEVELOPMENT_TEAM') == true) {
+      printError(
+        '═══════════════════════════════════════════════════════════════════════════════════');
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/ios/mac.dart]: I don't think wrapping here is helping readability.

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter_tools/lib/src/ios/mac.dart]: yeah that might be more readable.

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 8]
```diff
@@ -206,15 +213,57 @@ void diagnoseXcodeBuildFailure(XcodeBuildResult result) {
...
-  XcodeBuildResult(this.success, {this.output, this.stdout, this.stderr});
+  XcodeBuildResult(
+    this.success,
+    {
+      this.output,
+      this.stdout,
+      this.stderr,
+      this.buildCommands,
+      this.buildForDevice,
+      this.appDirectory
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/ios/mac.dart]: trailing comma

**Comment by Hixie** [comment]
LGTM but I don't really know much about Xcode.

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 3 above)
Comment (xster) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Some grouping would indeed be nicer. Pulled out an inner class and used assertions.

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 2 above)
Comment (xster) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Done

File: packages/flutter_tools/lib/src/commands/build_ios.dart (hunk 4 above)
Comment (xster) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: Swapped in the message 2 lines earlier. Makes sense to print the definitive implication at the end anyway.

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 5 above)
Comment (xster) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Done

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 6 above)
Comment (xster) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Done

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 8 above)
Comment (xster) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Done

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 1 above)
Comment (xster) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Check both development team and provisioning profile as suggested

**Pull request merged**